cosmoz-default-tree.html

```diff
@@ -23,134 +35,146 @@
 
 	Cosmoz.DefaultTree.prototype = Object.create(Cosmoz.Tree.prototype);
 
-	Cosmoz.DefaultTree.prototype.getNodeByProperty = function (propertyName, propertyValue) {
+	/**
+	 * Returns the first found Node.
+	 */
+	Cosmoz.DefaultTree.prototype.getNodeByProperty = function (propertyName, propertyValue, exact, nodes) {
 		if (propertyName === undefined || propertyValue === undefined) {
 			return;
 		}
+		// Defaults
+		exact = exact !== undefined ? exact : true;
 
 		if (propertyName === 'pathLocator') {
-			return this._getNodeByPathLocator(propertyValue);
+			return this.getNodeByPathLocator(propertyValue);
 		}
 
-		return this._searchAllRoots(propertyName, propertyValue);
+		var results = this.searchNodes(propertyName, propertyValue, exact, nodes);
+		return results && results.length > 0 ? results[0] : null;
 	};
 
-	Cosmoz.DefaultTree.prototype.getPath = function (node) {
-		if (node === undefined) {
-			return;
-		}
-		return this._getPathByPathLocator(node.pathLocator);
+	/**
+	 * Loops over a node list (or roots if not defined) and returns the concatenated results of search().
+	 */
+	Cosmoz.DefaultTree.prototype.searchNodes = function (propertyName, propertyValue, exact, nodes) {
```

Comment:
I'm a bit concerned that the behavior has changed completely from returning the first hit, to returning first or all, to always returning all. It also has a significant performance hit since we often expect only (or care about) just the first hit.
I also don't want to have too many arguments to methods, it's generally a sign to refactor.

How about:
```js
_searchNodes = function (propertyValue, options, nodes) { }
```
where the result gathering is something like:
```js
nodes.some(function (node) {
 	results = results.concat(this.search(node, propertyValue, options));
	return options.firstHitOnly && results.length > 0;
 }, this);
```
and `search()` is adapted as well.

Then we can expose two different methods to use it:

```js
searchNodes = function (propertyValue, exact, propertyName, nodes) {
	var options = {
		propertyName: propertyName || this.searchProperty,
		exact: exact !== undefined ? exact : true,
		firstHitOnly: false
	};
	return this._searchNodes(propertyValue, options, nodes);
}
findNode = function (propertyValue, propertyName, nodes) {
	var options = {
			propertyName: propertyName || this.searchProperty,
			exact: true,
			firstHitOnly: true
		},
		nodes = this._searchNodes(propertyValue, options, nodes);
	return nodes.shift() || null;
}
```

I think this would make the calling code more readable as well, where we more explicitly describe our intent to find a unique node, or to find nodes matching a pattern.

I think since `Cosmoz.DefaultTree` is an implementation, it should own the (default) contract of the tree and `propertyName` should actually be a property of the implementation - `this.searchProperty` - optionally configurable by the caller.

What do you think?